samples/nrf9160/aws_fota/src/main.c

```diff
@@ -0,0 +1,571 @@
+			err = mqtt_disconnect(c);
+			if (err) {
+				printk("Could not disconnect: %d\n", err);
+			}
+		}
+		break;
+
+	case MQTT_EVT_DISCONNECT:
+		printk("[%s:%d] MQTT client disconnected %d\n", __func__,
+		       __LINE__, evt->result);
+		break;
+
+	case MQTT_EVT_PUBLISH: {
+		const struct mqtt_publish_param *p = &evt->param.publish;
+
+		printk("[%s:%d] MQTT PUBLISH result=%d len=%d\n", __func__,
+		       __LINE__, evt->result, p->message.payload.len);
+		err = publish_get_payload(c, p->message.payload.len);
+		if (err) {
+			printk("mqtt_read_publish_payload: Failed! %d\n", err);
+			printk("Disconnecting MQTT client...\n");
+
+			err = mqtt_disconnect(c);
+			if (err) {
+				printk("Could not disconnect: %d\n", err);
+			}
+		}
+
+		if (p->message.topic.qos == MQTT_QOS_1_AT_LEAST_ONCE) {
+			const struct mqtt_puback_param ack = {
+				.message_id = p->message_id
+			};
+
+			/* Send acknowledgment. */
+			err = mqtt_publish_qos1_ack(c, &ack);
+			if (err) {
+				printk("unable to ack\n");
+			}
+		}
+		data_print("Received: ", payload_buf, p->message.payload.len);
+		break;
+	}
+
+	case MQTT_EVT_PUBACK:
+		if (evt->result != 0) {
+			printk("MQTT PUBACK error %d\n", evt->result);
+			break;
+		}
+
+		printk("[%s:%d] PUBACK packet id: %u\n", __func__, __LINE__,
+		       evt->param.puback.message_id);
+		break;
+
+	case MQTT_EVT_SUBACK:
+		if (evt->result != 0) {
+			printk("MQTT SUBACK error %d\n", evt->result);
+			break;
+		}
+
+		printk("[%s:%d] SUBACK packet id: %u\n", __func__, __LINE__,
+		       evt->param.suback.message_id);
+		break;
+
+	default:
+		printk("[%s:%d] default: %d\n", __func__, __LINE__,
+		       evt->type);
+		break;
+	}
+}
+
+
+/**@brief Resolves the configured hostname and
+ * initializes the MQTT broker structure
+ */
+static void broker_init(void)
+{
+	int err;
+	struct addrinfo *result;
+	struct addrinfo *addr;
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM
+	};
+
+	err = getaddrinfo(CONFIG_AWS_HOSTNAME, NULL, &hints, &result);
+	if (err) {
+		printk("ERROR: getaddrinfo failed %d\n", err);
+
+		return;
+	}
+
+	addr = result;
+	err = -ENOENT;
+
+	/* Look for address of the broker. */
+	while (addr != NULL) {
+		/* IPv4 Address. */
+		if (addr->ai_addrlen == sizeof(struct sockaddr_in)) {
+			struct sockaddr_in *broker4 =
+				((struct sockaddr_in *)&broker);
+
+			broker4->sin_addr.s_addr =
+				((struct sockaddr_in *)addr->ai_addr)
+				->sin_addr.s_addr;
+			broker4->sin_family = AF_INET;
+			broker4->sin_port = htons(8883);
+			printk("IPv4 Address found 0x%08x\n",
+			       broker4->sin_addr.s_addr);
+			break;
+		} else {
+			printk("ai_addrlen = %u should be %u or %u\n",
+			       (unsigned int)addr->ai_addrlen,
+			       (unsigned int)sizeof(struct sockaddr_in),
+			       (unsigned int)sizeof(struct sockaddr_in6));
+		}
+
+		addr = addr->ai_next;
+		break;
+	}
+
+	/* Free the address. */
+	freeaddrinfo(result);
+}
+
+static int provision(void)
```

Comment:
It's a bit weird to have `provision` and `provision_modem_key`.

Docs would be welcome as well.

Reply:
I agree this should be straightened up. 

I guess there are two approaches here because unfortunately, we do not have a common interface to provide certificates - Zephyr's internal stack uses `tls_credential_add` while nrf9160 has `nrf_inbuilt_key_write`. The question is whether we support QEMU as a target platform now - if not, I opt to drop the "native" approach.

Reply:
@rlubos maybe you could answer me because I need a confirmation of something. I want the default behavior to be.

Associate board with `nrf_cloud` through use of `asset_tracker` sample etc. Now the certs should be on the board if I understand this correctly. So I actually only need to provide the default SEC_TAG used by `asset_tracker` as the modem should find the certs there right?

Reply:
Yes, you are correct - `nrf_inbuilt_key_write` should store the certificate in persistent storage of the modem, so it should be possible to use the same SEC_TAG from a different app w/o a need to store them again.

Reply:
Tested and it works :)